pkg/services/serviceaccounts/proxy/service.go

```diff
@@ -0,0 +1,109 @@
+
+func (s *ServiceAccountsProxy) CreateServiceAccount(ctx context.Context, orgID int64, saForm *serviceaccounts.CreateServiceAccountForm) (*serviceaccounts.ServiceAccountDTO, error) {
+	if isNameValid(saForm.Name) {
+		s.log.Info("Unable to create service account with a protected name", "name", saForm.Name)
+		return nil, extsvcaccounts.ErrInvalidName
+	}
+	return s.proxiedService.CreateServiceAccount(ctx, orgID, saForm)
+}
+
+func (s *ServiceAccountsProxy) DeleteServiceAccount(ctx context.Context, orgID, serviceAccountID int64) error {
+	sa, err := s.proxiedService.RetrieveServiceAccount(ctx, orgID, serviceAccountID)
+	if err != nil {
+		return err
+	}
+
+	if isExternalServiceAccount(sa.Login) {
+		s.log.Info("unable to delete external service accounts", "serviceAccountID", serviceAccountID)
+		return extsvcaccounts.ErrCannotBeDeleted
+	}
+
+	return s.proxiedService.DeleteServiceAccount(ctx, orgID, serviceAccountID)
+}
+
+func (s *ServiceAccountsProxy) RetrieveServiceAccount(ctx context.Context, orgID, serviceAccountID int64) (*serviceaccounts.ServiceAccountProfileDTO, error) {
+	sa, err := s.proxiedService.RetrieveServiceAccount(ctx, orgID, serviceAccountID)
+	if err != nil {
+		return nil, err
+	}
+
+	sa.IsExternal = isExternalServiceAccount(sa.Login)
+
+	return sa, nil
+}
+
+func (s *ServiceAccountsProxy) RetrieveServiceAccountIdByName(ctx context.Context, orgID int64, name string) (int64, error) {
+	return s.proxiedService.RetrieveServiceAccountIdByName(ctx, orgID, name)
+}
+
+func (s *ServiceAccountsProxy) UpdateServiceAccount(ctx context.Context, orgID, serviceAccountID int64, saForm *serviceaccounts.UpdateServiceAccountForm) (*serviceaccounts.ServiceAccountProfileDTO, error) {
+	if isExternalServiceAccount(*saForm.Name) {
+		s.log.Info("Invalid service account name", "name", *saForm.Name)
+		return nil, extsvcaccounts.ErrInvalidName
+	}
+	sa, err := s.proxiedService.RetrieveServiceAccount(ctx, orgID, serviceAccountID)
+	if err != nil {
+		return nil, err
+	}
+	if isExternalServiceAccount(sa.Login) {
+		s.log.Info("unable to update external service accounts", "serviceAccountID", serviceAccountID)
+		return nil, extsvcaccounts.ErrCannotBeUpdated
+	}
+
+	return s.proxiedService.UpdateServiceAccount(ctx, orgID, serviceAccountID, saForm)
+}
+
+func (s *ServiceAccountsProxy) AddServiceAccountToken(ctx context.Context, serviceAccountID int64, cmd *serviceaccounts.AddServiceAccountTokenCommand) (*apikey.APIKey, error) {
+	sa, err := s.proxiedService.RetrieveServiceAccount(ctx, cmd.OrgId, serviceAccountID)
+	if err != nil {
+		return nil, err
+	}
+
+	if isExternalServiceAccount(sa.Login) {
+		s.log.Info("unable to create tokens for external service accounts", "serviceAccountID", serviceAccountID)
+		return nil, extsvcaccounts.ErrCannotCreateToken
+	}
+
+	return s.proxiedService.AddServiceAccountToken(ctx, serviceAccountID, cmd)
+}
+
+func isNameValid(name string) bool {
+	return strings.HasPrefix(name, extsvcaccounts.ExtsvcPrefix)
+}
```

Comment:
I think the logic needs to be reversed here
```suggestion
func isNameValid(name string) bool {
	return !strings.HasPrefix(name, extsvcaccounts.ExtsvcPrefix)
}
```

Reply:
changed